libs/designer/src/lib/core/state/undoRedo/undoRedoTypes.ts

```diff
@@ -0,0 +1,62 @@
+import { updateNodeConnection } from '../../actions/bjsworkflow/connections';
+import type { RootState } from '../../store';
+import {
+  addEdgeFromRunAfter,
+  addImplicitForeachNode,
+  addNode,
+  addSwitchCase,
+  deleteNode,
+  deleteSwitchCase,
+  moveNode,
+  pasteNode,
+  pasteScopeNode,
+  removeEdgeFromRunAfter,
+  updateRunAfter,
+} from '../workflow/workflowSlice';
+
+export interface StateHistory {
+  past: string[];
+  future: string[];
+}
+
+// Omitted slices: designerView, designerOptions, dev, undoRedo
+export type UndoRedoPartialRootState = Pick<
+  RootState,
+  | 'connections'
+  | 'customCode'
+  | 'operations'
+  | 'panel'
+  | 'panelV2'
+  | 'settings'
+  | 'staticResults'
+  | 'tokens'
+  | 'workflow'
+  | 'workflowParameters'
+>;
+
+export const undoableActionTypes = [
+  addNode,
+  moveNode,
+  deleteNode,
+  addSwitchCase,
+  deleteSwitchCase,
+  addImplicitForeachNode,
+  pasteNode,
+  pasteScopeNode,
+  updateRunAfter,
+  removeEdgeFromRunAfter,
+  addEdgeFromRunAfter,
+  updateNodeConnection.fulfilled,
+  /*
+   * TODO: Following actions to be added after bug fixes for:
+   * 1. Parameters panel is not re-rendering on undo/redo click
+   * 2. State history should be saved only when there is a change in state (clicking in/out of parameter is triggering save)
+   *
+   * Actions list:
```

Comment:
I think we should combine all of the operation parameter / settings type actions as one state save that runs when the panel is closed if there were any edits.  We can do this in a separate PR but I think the flow should go:

1. Open panel
2. User makes edits
3. We mark the state as "dirty"
4. User tries to leave panel / we are "done" with the edits
5. Save state in history
6. Close panel

This would avoid issue #2

Reply:
Ohh, I'll check with our PM on whether this is ok. In V1, we were storing a state for every parameter change (i.e. when user focused out of a parameter and there was a change in flow definition). I was thinking of looking into why the re-render isn't happening when panel is open on undo/redo click; and to add a condition on save state function to save only if there was a change from previous state.

Reply:
Ah ok I wasn't aware you had undo/redo previously, we can keep that same functionality then that's fine